### PR TITLE
Installers: Add Base Installer Files Ahead Of JDK21 release.

### LINKED ISSUE #736: Add installer files for JDK21
Add installer files for JDK21

### PR DESCRIPTION
**Pull request by steelhead31**
DO NOT MERGE UNTIL CHECKSUMS ARE CORRECT  
  
Add the initial base files which will require the usual filename, checksum and version number updates during the release.

Fixes #736

### CODE REVIEW

**Comment by steelhead31** [comment]
Checks fail, as the binary files are not present yet.

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	libxext-dev
+	libxrandr-dev
+	libxrender-dev
+	libxt-dev
+	libxtst-dev
+"
+depends=""
+subpackages="$pkgname-src:_src:noarch
+	$pkgname-jdk:_jdk:x86_64"
+source="https://github.com/adoptium/temurin21-binaries/releases/download/jdk-$_pkgvername/OpenJDK21U-jdk_x64_alpine-linux_hotspot_$_pkgver.tar.gz
```
Comment (adamfarley) [inline linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD:29]: Is the "U" correct, prior to October?

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	HelloWorld.java
+	TestECDSA.java
+	TestCryptoLevel.java
+"
+
+_java_home="/usr/lib/jvm/java-19-temurin"
```
Comment (adamfarley) [inline linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD:36]: Should this be jdk20?

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	mv "$_fromroot/NOTICE"  "$_toroot"
+
+	# symlink to shared cacerts store
+	rm "$_toroot/lib/security/cacerts"
+	ln -sf /etc/ssl/certs/java/cacerts \
+		"$_toroot/lib/security/cacerts"
+}
+
+sha256sums="
+b03aced4b7a1c49bc00297e35e45480fd03818862b93e17e1551a3b721e89306  OpenJDK21U-jdk_x64_alpine-linux_hotspot_$_pkgver.tar.gz
```
Comment (adamfarley) [inline linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD:98]: Another "U".

File: linux/jdk/debian/src/main/packaging/temurin/21/debian/rules [hunk 4]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/make -f
+
+pkg_name = temurin-21-jdk
+priority = 2100
+jvm_tools = jar jarsigner java javac javadoc javap jcmd jconsole jdb jdeprscan jdeps jfr jhsdb jimage jinfo jlink jmap jmod jpackage jps jrunscript jshell jstack jstat jstatd jwebserver keytool rmiregistry serialver jexec jspawnhelper
+amd64_tarball_url = https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.0%2B35/OpenJDK21U-jdk_x64_linux_hotspot_21.0.0_35.tar.gz
```
Comment (adamfarley) [inline linux/jdk/debian/src/main/packaging/temurin/21/debian/rules:6]: "U"

File: linux/jdk/debian/src/main/packaging/temurin/21/debian/rules [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+amd64_checksum = 3d91842e9c172967ac397076523249d05a82ead51b0006838f5f0315ad52222c
+arm64_tarball_url = https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.0%2B35/OpenJDK21U-jdk_aarch64_linux_hotspot_21.0.0_35.tar.gz
```
Comment (adamfarley) [inline linux/jdk/debian/src/main/packaging/temurin/21/debian/rules:8]: "U"

File: linux/jdk/suse/src/main/packaging/temurin/21/temurin-21-jdk.spec [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+Provides: java-%{java_provides}-devel
+Provides: java-%{java_provides}-headless
+Provides: java-sdk
+Provides: java-sdk-21
+Provides: java-sdk-21-%{java_provides}
+Provides: java-sdk-%{java_provides}
+
+# First architecture (x86_64)
+Source0: %{source_url_base}/jdk-%{upstream_version_url}/OpenJDK21U-jdk_%{vers_arch}_linux_hotspot_%{upstream_version_no_plus}.tar.gz
+Source1: %{source_url_base}/jdk-%{upstream_version_url}/OpenJDK21U-jdk_%{vers_arch}_linux_hotspot_%{upstream_version_no_plus}.tar.gz.sha256.txt
```
Comment (adamfarley) [inline linux/jdk/suse/src/main/packaging/temurin/21/temurin-21-jdk.spec:100]: "U"

Also, I'm going to stop tagging all the "U"s. There's a bunch more of them, and the million one-letter-comments doesn't add value.

**Review by adamfarley — CHANGES_REQUESTED** [review]
A few points to consider. Once addressed (fixed/justified), will approve.

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD (hunk 2 above)
Comment (steelhead31) [reply linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD]: Yup, fixed!

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD (hunk 3 above)
Comment (steelhead31) [reply linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD]: Yup, fixed!

File: linux/jdk/debian/src/main/packaging/temurin/21/debian/rules (hunk 5 above)
Comment (steelhead31) [reply linux/jdk/debian/src/main/packaging/temurin/21/debian/rules]: Yup, fixed!

File: linux/jdk/suse/src/main/packaging/temurin/21/temurin-21-jdk.spec [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+Provides: java-%{java_provides}
+Provides: java-%{java_provides}-devel
+Provides: java-%{java_provides}-headless
+Provides: java-sdk
+Provides: java-sdk-21
+Provides: java-sdk-21-%{java_provides}
+Provides: java-sdk-%{java_provides}
+
+# First architecture (x86_64)
+Source0: %{source_url_base}/jdk-%{upstream_version_url}/OpenJDK21U-jdk_%{vers_arch}_linux_hotspot_%{upstream_version_no_plus}.tar.gz
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/jdk/suse/src/main/packaging/temurin/21/temurin-21-jdk.spec]: Yup, fixed!

**Review by adamfarley — APPROVED** [review]
LGTM :)

**Review by karianna — CHANGES_REQUESTED** [review]

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+subpackages="$pkgname-src:_src:noarch
+	$pkgname-jdk:_jdk:x86_64"
+source="https://github.com/adoptium/temurin21-binaries/releases/download/jdk-$_pkgvername/OpenJDK21-jdk_x64_alpine-linux_hotspot_$_pkgver.tar.gz
+
+	HelloWorld.java
+	TestECDSA.java
+	TestCryptoLevel.java
+"
+
+_java_home="/usr/lib/jvm/java-20-temurin"
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD]: Fixed

File: linux/Jenkinsfile [hunk 9]
```diff
@@ -36,7 +36,7 @@ pipeline {
     }
     parameters {
         choice(name: 'TYPE', choices: Types.ALL, description: 'Build JDK or JRE')
-        choice(name: 'VERSION', choices: ['8', '11', '17', '20'], description: 'Build for specific JDK VERSION')
+        choice(name: 'VERSION', choices: ['8', '11', '17', '20', '21'], description: 'Build for specific JDK VERSION')
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/Jenkinsfile]: Removed JDK20 :)

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD [hunk 10]
```diff
@@ -0,0 +1,102 @@
+# Maintainer: Eclipse Adoptium Package Maintainers <temurin-dev@eclipse.org>
+pkgname=temurin-21
+pkgver=21.0.0_p35
+# replace _p1 with _1
+_pkgver=${pkgver/_p/_}
+_pkgverplus=${pkgver/_p/+}
+_pkgvername=${_pkgverplus/+/%2B}
+pkgrel=0
+pkgdesc="Eclipse Temurin 21"
+provider_priority=21
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD:10]: I think its correct, all of the previous versions of the APKBUILD file have the provider priority set as the JDK version, @gdams can you confirm ?

File: linux/jdk/debian/src/main/packaging/temurin/21/debian/rules [hunk 11]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/make -f
+
+pkg_name = temurin-21-jdk
+priority = 2100
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/jdk/debian/src/main/packaging/temurin/21/debian/rules:4]: I've updated this to match the previous debian version, ie 1112 for JDK11, 1712 for JDK17 etc.

File: linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD (hunk 10 above)
[the remark this answers is not in the document]
Comment (gdams) [reply linux/jdk/alpine/src/main/packaging/temurin/21/APKBUILD:10]: Yeah looking at how the community do this for Alpine we are doing the same which makes sense https://git.alpinelinux.org/aports/tree/community/openjdk17/APKBUILD?h=3.15-stable

**Review by karianna — APPROVED** [review]

**Comment by steelhead31** [comment]
DO NOT MERGE UNTIL CHECKSUMS ARE CORRECTED FOR RELEASE

File: linux/jdk/redhat/src/main/packaging/temurin/21/temurin-21-jdk.spec [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    update-alternatives --remove javac %{prefix}/bin/javac
+fi
+
+%files
+%defattr(-,root,root)
+%{prefix}
+/usr/lib/tmpfiles.d/%{name}.conf
+
+%changelog
+* Wed Sep 20 2023 Eclipse Adoptium Package Maintainers <temurin-dev@eclipse.org> 21.0.0.0.0.35-0
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/jdk/redhat/src/main/packaging/temurin/21/temurin-21-jdk.spec]: Fixed, thanks!

**Comment by gdams** [comment]
@steelhead31 we've started publishing the binaries now. Can you update the shasums?

File: linux/jdk/debian/src/main/packaging/temurin/21/debian/control [hunk 13]
```diff
@@ -0,0 +1,99 @@
+Source: temurin-21-jdk
+Section: java
+Priority: optional
+Maintainer: Eclipse Adoptium Package Maintainers <temurin-dev@eclipse.org>
+Build-Depends: debhelper (>= 11), lsb-release
+
+Package: temurin-21-jdk
+Architecture: amd64 armhf arm64 ppc64el s390x
```
Comment (gdams) [inline linux/jdk/debian/src/main/packaging/temurin/21/debian/control:8]:
```suggestion
Architecture: amd64 arm64 ppc64el
```

**Review by jiekang — APPROVED** [review]
Specs look fine to me. Would be nice if the GH check could test them. Wonder if it's worthwhile in making the specs usable when not all arches are available... but for a separate conversation

File: linux/jre/debian/src/main/packaging/temurin/21/debian/control [hunk 14]
```diff
@@ -0,0 +1,63 @@
+Source: temurin-21-jre
+Section: java
+Priority: optional
+Maintainer: Eclipse Adoptium Package Maintainers <temurin-dev@eclipse.org>
+Build-Depends: debhelper (>= 11), lsb-release
+
+Package: temurin-21-jre
+Architecture: amd64 armhf arm64 ppc64el s390x
```
Comment (gdams) [inline linux/jre/debian/src/main/packaging/temurin/21/debian/control:8]:
```suggestion
Architecture: amd64 arm64 ppc64el
```

**Review by gdams — APPROVED** [review]
LGTM (subject to ppc64le either being removed or waiting until it's been shipped)

**Pull request merged**